include/mptcpd/path_manager_private.h

```diff
@@ -67,6 +78,212 @@ struct mptcpd_pm
         struct mptcpd_nm *nm;
 };
 
+/**
+ * @struct mptcpd_pm_cmd_ops
+ *
+ * @brief MPTCP path management generic netlink command functions.
+ *
+ * The set of functions that implement MPTCP path management generic
+ * netlink command calls.
+ */
+struct mptcpd_pm_cmd_ops
+{
+        /**
+         * @privatesection
+         */
+        /**
+         * @name Common Path Management Commands
+         *
+         * Path management common to both the upstream and
+         * multipath-tcp.org Linux kernels.
+         *
+         * @param[in] pm    The mptcpd path manager object.
+         * @param[in] addr  Local IP address and port to be advertised
+         *                  through the MPTCP protocol @c ADD_ADDR
+         *                  option.  The port is optional, and is
+         *                  ignored if it is zero.
+         * @param[in] id    MPTCP local address ID.
+         * @param[in] flags
+         * @param[in] index Network interface index (optional for
+         *                  upstream Linux kernel).
+         * @param[in] token MPTCP connection token.
+         */
+        //@{
+        /**
+         * @brief Advertise new network address to peers.
+         */
+        int (*add_addr)(struct mptcpd_pm *pm,
+                        struct sockaddr const *addr,
+                        mptcpd_aid_t id,
+                        uint32_t flags,
+                        int index,
+                        mptcpd_token_t token);
+
+        /**
+         * @brief Stop advertising network address to peers.
+         */
+        int (*remove_addr)(struct mptcpd_pm *pm,
+                           mptcpd_aid_t address_id,
+                           mptcpd_token_t token);
+        //@}
+
+        /**
+         * @name Upstream Kernel Path Management Commands
+         *
+         * Path management commands supported by the upstream Linux
+         * kernel.
+         */
+        //@{
+        /**
+         * @brief Get network address corresponding to an address ID.
+         *
+         * @param[in]  pm   The mptcpd path manager object.
+         * @param[in]  id   MPTCP local address ID.
+         * @param[out] addr Network address information associated
+         *                  with the @a id.  Deallocate with
+         *                  @c mptcpd_addr_info_destroy().
+         *
+         * @return @c 0 if operation was successful. -1 or @c errno
+         *         otherwise.
+         */
+        int (*get_addr)(struct mptcpd_pm *pm,
+                        mptcpd_aid_t id,
+                        struct mptcpd_addr_info **addr);
+
+        /**
+         * @brief Dump list of network addresses.
+         *
+         * @param[in]  pm   The mptcpd path manager object.
+         * @param[out] addrs Array of network addresses.  Deallocate
+         *                   by iterating over the array, and calling
+         *                   @c mptcpd_addr_info_destroy() on each
+         *                   element.
```

Comment:
There is no longer a need for the user to destroy `mptcpd_addr_info` objects or arrays since they are now passed as "in" arguments to the newly introduced user callbacks.  Mptcpd owns the memory, and deallocates after returning from the user callback call.